neqo-http3/src/lib.rs

```diff
@@ -184,6 +184,59 @@ use thiserror::Error;
 
 use crate::{features::extended_connect, priority::PriorityHandler};
 
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub struct DatagramPayload {
+    data: Vec<u8>,
+    payload_offset: usize,
+}
+
+impl DatagramPayload {
+    /// # Panics
+    ///
+    /// If the payload offset lies outside the data.
+    #[must_use]
+    pub fn new(data: Vec<u8>, payload_offset: usize) -> Self {
+        assert!(payload_offset <= data.len());
+        Self {
+            data,
+            payload_offset,
+        }
+    }
+
+    #[must_use]
+    pub fn payload(&self) -> &[u8] {
+        &self.data[self.payload_offset..]
+    }
+}
+
+impl AsRef<[u8]> for DatagramPayload {
+    fn as_ref(&self) -> &[u8] {
+        self.payload()
+    }
+}
+
+impl From<DatagramPayload> for Vec<u8> {
```

Comment:
Want to implement a `From` in the other direction as well?  A `Vec<u8>` is trivially a `DatagramPayload` with zero offset.